From df3f339d2637b2aa8d9e682c5eb08b26dda3dc36 Mon Sep 17 00:00:00 2001
From: Yaniv Rosner <yanivr@broadcom.com>
Date: Wed, 27 Feb 2013 13:06:45 +0000
Subject: [PATCH 605/858] bnx2x: Fix KR2 link

Fix KR2 link down problem after reboot when link speed is reconfigured via ethtool.
Since 1G/10G support link speed were missing by default, 1G/10G link speed were
not advertised.

Signed-off-by: Yaniv Rosner <yanivr@broadcom.com>
Signed-off-by: Eilon Greenstein <eilong@broadcom.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit be94bea753a4d5ac0982df07aaeaf0cb1f7554dd)

Signed-off-by: Arnaldo Carvalho de Melo <acme@redhat.com>
---
 drivers/net/ethernet/broadcom/bnx2x/bnx2x_ethtool.c | 6 ++++++
 drivers/net/ethernet/broadcom/bnx2x/bnx2x_link.c    | 2 ++
 2 files changed, 8 insertions(+)

diff --git a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_ethtool.c b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_ethtool.c
index bd5dca1..f3f5acc 100644
--- a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_ethtool.c
+++ b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_ethtool.c
@@ -281,6 +281,8 @@ static int bnx2x_get_settings(struct net_device *dev, struct ethtool_cmd *cmd)
    cmd->lp_advertising |= ADVERTISED_2500baseX_Full;
   if (status & LINK_STATUS_LINK_PARTNER_10GXFD_CAPABLE)
    cmd->lp_advertising |= ADVERTISED_10000baseT_Full;
+  if (status & LINK_STATUS_LINK_PARTNER_20GXFD_CAPABLE)
+   cmd->lp_advertising |= ADVERTISED_20000baseKR2_Full;
  }
 
  cmd->maxtxpkt = 0;
@@ -463,6 +465,10 @@ static int bnx2x_set_settings(struct net_device *dev, struct ethtool_cmd *cmd)
       ADVERTISED_10000baseKR_Full))
     bp->link_params.speed_cap_mask[cfg_idx] |=
      PORT_HW_CFG_SPEED_CAPABILITY_D0_10G;
+
+   if (cmd->advertising & ADVERTISED_20000baseKR2_Full)
+    bp->link_params.speed_cap_mask[cfg_idx] |=
+     PORT_HW_CFG_SPEED_CAPABILITY_D0_20G;
   }
  } else { /* forced speed */
   /* advertise the requested speed and duplex if supported */
diff --git a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_link.c b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_link.c
index e68545a..c53ab4c 100644
--- a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_link.c
+++ b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_link.c
@@ -11851,6 +11851,8 @@ static int bnx2x_populate_int_phy(struct bnx2x *bp, u32 shmem_base, u8 port,
    phy->media_type = ETH_PHY_KR;
    phy->flags |= FLAGS_WC_DUAL_MODE;
    phy->supported &= (SUPPORTED_20000baseKR2_Full |
+        SUPPORTED_10000baseT_Full |
+        SUPPORTED_1000baseT_Full |
         SUPPORTED_Autoneg |
         SUPPORTED_FIBRE |
         SUPPORTED_Pause |
-- 
1.8.3.1